src/main/java/SYSC4806/Controller/HomeController.java

```diff
@@ -40,27 +43,73 @@ public String home(Model model) {
     }
 
     /*
-    * Selects courses based on chosen program and year and returns them to view
-    * to refresh results table
+     * Selects courses based on chosen program, year, courses and catagories and returns them to view
+     * to refresh results table
      */
     @GetMapping("/results/{program}/{year}")
-    public String showResults(Model model, @PathVariable("program") String p, @PathVariable String year, @RequestParam(value = "courses", required = false) String co)
-    {
+    public String showResults(Model model, @PathVariable("program") String p, @PathVariable String year,
+                              @RequestParam(value = "courses", required = false) String co,
+                              @RequestParam(value = "categories", required = false) String ca) {
+
         long p_id = programRepository.findByName(p).getId();
         List<String> courseNames = courseRepository.findCourseByProgramAndYear(p_id, Integer.parseInt(year));
 
-        //Edge case: filter doesnt come up with something
-        if(co!=null){
+        if (co != null && !(co.isEmpty())) {
             List<String> courseNamesFilter = Arrays.asList(co.split("\\s*,\\s*"));
             courseNames.retainAll(courseNamesFilter);
         }
 
+        if (ca != null && !(ca.isEmpty())) {
+            List<String> catNamesFilter = Arrays.asList(ca.split("\\s*,\\s*"));
+
+            //As per the category, finding the list of learning outcomes
+            ArrayList<Long> loListId = new ArrayList<>();
+            for (String cName : catNamesFilter) {
+                Category category = categoryRepository.findByName(cName);
+                loListId.add(learningOutcomeRepository.findByCategoryId(category.getId()).getId());
+            }
+
+            //As per the list of learning outcomes, find the list of course id's
+            ArrayList<BigInteger> courseIdsList = new ArrayList<>();
+            for (Long lo : loListId) {
+                List<BigInteger> courseIds = learningOutcomeRepository.findByLearningOutcomeId(lo);
+                courseIdsList.addAll(courseIds);
+            }
+
+            //As per course ids, find the course names
+            ArrayList<BigInteger> courseIdsListUnique = removeDuplicates(courseIdsList); //remove duplicates
```

Comment:
Would it be more efficient to implement this as a HashSet? So when you're adding to the courseIdsList, any duplicates are automatically not added if it already exists within the set.